src/calibre/utils/icu_calibre_utils.h

```diff
@@ -114,3 +115,110 @@ static PyObject* icu_to_python(UChar *src, int32_t sz) {
 #endif
 }
 #endif
+
+#else  // end PY2; start PY3.3+
+
+static UChar* python_to_icu(PyObject *obj, int32_t *osz) {
+    UChar *ans = NULL;
+    Py_ssize_t sz = 0;
+    UErrorCode status = U_ZERO_ERROR;
+    int i;
+
+    if (!PyUnicode_CheckExact(obj)) {
+        PyErr_SetString(PyExc_TypeError, "Not a unicode string");
+        return NULL;
+    }
+    if(PyUnicode_READY(obj) == -1) {
+        return NULL;
+    }
+    sz = PyUnicode_GET_LENGTH(obj);
+
+
+    switch(PyUnicode_KIND(obj)) {
+    case PyUnicode_1BYTE_KIND:
+        ans = (UChar*) malloc((sz+1) * sizeof(UChar));
+        if (ans == NULL) {
+            PyErr_NoMemory();
+            return NULL;
+        }
+        u_strFromUTF8(
+            ans, sz + 1,
+            (int32_t*) osz,
+            (char*) PyUnicode_1BYTE_DATA(obj),
+            (int32_t) sz,
+            &status);
+        break;
+    case PyUnicode_2BYTE_KIND:
+        ans = (UChar*) malloc((sz+1) * sizeof(UChar));
```

Comment:
cant we use sizeof(UChar) to check?

Reply:
Yes, you're right. I even looked at the assembly, and was surprised to see that GCC-4.8 doesn't automatically switch to memcpy (although it does use a unrolled loop)